### PR TITLE
Issue 38 - adding unit tests

### PR DESCRIPTION
**Pull request by Holdrick**
Contributes to #38  (add issue number here. i.e. `#302`)

## What did you do?
- Added unit tests to the analyse_text and check_words methods in main.py
- Extracted db creation to a method that's called in each required method so the db creation can be mocked out in unit tests
## Why did you do it?
Adding unit tests helps protect us from future changes that break previous ones
## How have you tested it?
I did some manual tests to see if my change broke anything
## Were docs updated if needed?

- [ ] No
- [ ] Yes
- [x] N/A

#### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new console logs
- [ ] Fixes entire issue
- [x] Partial fix for issue

### CODE REVIEW

**Comment by Holdrick** [comment]
Looks like theres a conflict with the old test. The test was written before token authentication so I reworked it to pass with the present production code

File: taketwo-webapi/main.py [hunk 1]
```diff
@@ -97,11 +96,12 @@ def validate_token_IBM(token, authURL, clientId, clientSecret=Depends(oauth2_sch
 
 
 client = couchdb.Server(f'http://{db_username}:{db_password}@{db_host}:{db_port}/')
-try: 
-    db = client.create(db_name)
-except couchdb.PreconditionFailed:
-    db = client[db_name]
 
+def getDb():
```
Comment (Holdrick) [inline taketwo-webapi/main.py:100]: Moved the db creation to a method so it can be mocked in unit tests

File: taketwo-webapi/main.py [hunk 2]
```diff
@@ -217,9 +222,10 @@ def analyse_text(text: Text):
 @app.put("/check")
 def check_words(text: Text):
     res = []
+    db = getDb()
     for item in db.view('_all_docs',include_docs=True):
         doc = item.doc
-        if doc["category"] == "racial slur" and doc["flagged_string"].lower() in text.content.lower():
+        if doc["category"] == "racial-slur" and doc["flagged_string"].lower() in text.content.lower():
```
Comment (Holdrick) [inline taketwo-webapi/main.py:228]: Previously, this didn't match what was being returned from the read_categories method so the marks saved through template.html weren't recognized by this method

File: testing/integration/test.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    resp = client.post(url, headers=headers, data=json.dumps(payload,indent=4))       
+    
+    # Validate response headers and body contents, e.g. status code.
+    assert resp.status_code == 200, "Unexpected status code. Was: " + str(resp.status_code)
+    resp_body = resp.json()
+    assert resp_body['url'] == 'example.com', "Unexpected url. Was: " + resp_body['url']
+
+    #make sure payload stored properly
+    storedPayload = getDb().get(resp_body["_id"])
+    comparePayloads(payload, storedPayload)
```
Comment (Holdrick) [inline testing/integration/test.py:39]: Different from the previous version, add comparison of what's stored in the db to the payload we send to make sure it's stored properly

File: testing/integration/test.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
+import requests
+#example api test
+
+import json
+import sys
+sys.path.append('../../taketwo-webapi')
+sys.path.append('../util')
```
Comment (Holdrick) [inline testing/integration/test.py:7]: I'm new to python and dependencies so I'm not sure if this is optimal for being able to access adjacent files

File: testing/integration/test.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+from main import app
+from main import validate
+from main import getDb
+from fastapi.testclient import TestClient
+from util.assert_util import comparePayloads
+
+def override_validate():
+    return {"sub": 'test'}
+
+app.dependency_overrides[validate] = override_validate
```
Comment (Holdrick) [inline testing/integration/test.py:17]: Bypass validate so we don't have to connect to an off-system resource

File: testing/unit/test_analyse_text.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+import unittest
+from unittest.mock import patch
+
+from main import analyse_text
+from main import Text
+
+from util.mock_db_util import getRow
+from util.mock_db_util import setupMocks
+
+class TestAnalyseText(unittest.TestCase):
```
Comment (Holdrick) [inline testing/unit/test_analyse_text.py:14]: I'm not super familiar with python test coding standards but having the class name as the method under test and each method inside the class describing the test case seemed reasonable to me

File: testing/unit/util/mock_db_util.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
+from couchdb.client import Document
+from couchdb.client import Row
+from unittest.mock import MagicMock
+    
+def getRow(flaggedString, category, info):
```
Comment (Holdrick) [inline testing/unit/util/mock_db_util.py:5]: Moved these methods here so they can be used in both tests.

File: taketwo-webapi/main.py (hunk 2 above)
Comment (Holdrick) [reply taketwo-webapi/main.py:228]: I can take this out if we'd rather have this fixed in a separate change

**Comment by Holdrick** [comment]
The unit tests are mostly independent. There's only a couple things in main that could be refactored:
- You still need to be in the virtual environment
- The client variable aka the couchdb Server class initialised on line 98 (this seems okay for now. I've tested the unit tests without the docker container running)

**Review by upkarlidder — APPROVED** [review]

File: testing/integration/test.py (hunk 4 above)
Comment (Holdrick) [reply testing/integration/test.py:7]: Cool! I'll update

**Pull request merged**